modules/pico_6lowpan.c

```diff
@@ -0,0 +1,1652 @@
+ *  Global Variables
+ ******************************************************************************/
+
+static struct pico_queue pico_6lowpan_in = {
+    0
+};
+static struct pico_queue pico_6lowpan_out = {
+    0
+};
+
+static uint16_t dgram_tag = 0;
+
+/*******************************************************************************
+ *  Private functions
+ ******************************************************************************/
+
+/* Copies two memory buffers but also considers overlapping buffers */
+static void
+buf_move(void *dst, const void *src, size_t len)
+{
+    uint8_t *d = (uint8_t *)dst;
+    const uint8_t *s = (const uint8_t *)src;
+    if (!dst || !src) {
+        return;
+    } else {
+        if (d < s) {
+            while (len--)
+                *d++ = *s++;
+        } else {
+            s = s + len - 1;
+            d = d + len - 1;
+            while (len--)
+                *d-- = *s--;
+        }
+    }
+}
+
+/*******************************************************************************
+ *  Frags
+ ******************************************************************************/
+
+/* Compares two fragmentation cookies based on the hash */
+static int
+frag_ctx_cmp(void *a, void *b)
+{
+    struct frag_ctx *fa = (struct frag_ctx *)a;
+    struct frag_ctx *fb = (struct frag_ctx *)b;
+    return (int)(fa->hash - fb->hash);
+}
+
+/* Compares two fragmentation cookies according to RFC4944 5.3 */
+static int
+frag_cmp(void *a, void *b)
+{
+    struct frag_ctx *fa = (struct frag_ctx *)a;
+    struct frag_ctx *fb = (struct frag_ctx *)b;
+    int ret = 0;
+    if (fa->dgram_size != fb->dgram_size) {
+        return (int)(fa->dgram_size - fb->dgram_size);
+    } else if (fa->dgram_tag != fb->dgram_tag) {
+        return (int)(fa->dgram_tag - fb->dgram_tag);
+    } else {
+        ret = pico_6lowpan_ll_cmp(&fa->src, &fb->src, fa->f->dev);
+        if (ret)
+            return ret;
+        else {
+            return pico_6lowpan_ll_cmp(&fa->dst, &fb->dst, fa->f->dev);
+        }
+    }
+}
+
+PICO_TREE_DECLARE(FragTree, &frag_ctx_cmp);
+PICO_TREE_DECLARE(ReassemblyTree, &frag_cmp);
+
+/* Find a fragmentation cookie for transmission of subsequent fragments */
+static struct frag_ctx *
+frag_ctx_find(uint32_t hash)
+{
+    struct frag_ctx f = { .hash = hash };
+    return pico_tree_findKey(&FragTree, &f);
+}
+
+/* Reassembly timeout function, deletes */
+static void
+frag_timeout(pico_time now, void *arg)
+{
+    struct pico_tree_node *i = NULL, *next = NULL;
+    struct frag_ctx *key = NULL;
+    IGNORE_PARAMETER(arg);
+    pico_tree_foreach_safe(i, &ReassemblyTree, next) {
+        if ((key = i->keyValue)) {
+            if ((pico_time)(FRAG_TIMEOUT * 1000) <= (now - key->timestamp)) {
+                pico_tree_delete(&ReassemblyTree, key);
+                pico_frame_discard(key->f);
+                PICO_FREE(key);
+            }
+        }
+    }
+    pico_timer_add(1000, frag_timeout, NULL);
+}
```

Comment:
If it fails, the fragment should not be accepted and I also think the corresponding reassembly, if there is any going on for the fragment, should be flushed, since there's isn't any recovery available from these errors. Ack!